### ISSUE TITLE
Provide mechanism for hiding imports

### ISSUE THREAD

**Issue by jhance** [python/mypy#5736]
Imports are usually (not in all cases) an implementation detail, but it ends up getting "exported" unless you do
```
from foo import bar as _bar
```
to make it an error to import `bar` from the module. (Of course, one can still import `_bar`, but this is less likely to occur by accident)

In a large codebase, importing identifiers from the wrong module can cause unnecessary dependencies, which hurts selective testing. I've had to do the above hack in some cases where I've noticed that numerous files were incorrectly (possibly because the identifier got moved at some point) importing an identifier from a file that had a lot of dependencies as opposed to the one which was rather clean.

Another issue this can cause is it makes it impossible to safely and automatically remove unused exports, as its possible that they are being relied on somewhere.

A good start would be to allow `from import import bar as bar` in non-stubs (gated by an option). Even better would be an option to make `from import import bar` behave that way by default (and perhaps a way to mark identifiers as exported, but that isn't strictly needed since most files that re-export could just be have the flag disabled in mypy.ini)

**Comment by hauntsaninja** [comment]
We have this now `--no-implicit-reexport` is even part of strict
